enhancement(kafka source): Support Kafka headers in Kafka source

daniilguit opened the pull request:
* enhancement(kafka.rs): Added Kafka Headers support in Kafka source
  * resolves #6806

jszwedko commented on src/sources/kafka.rs:
```diff
@@ -375,5 +408,6 @@ mod integration_test {
         assert_eq!(events[0].as_log()["topic"], topic.into());
         assert!(events[0].as_log().contains("partition"));
         assert!(events[0].as_log().contains("offset"));
+        assert!(events[0].as_log().contains("headers"));
```
Could we assert on the actual values here too?

jszwedko commented on src/sources/kafka.rs:
```diff
@@ -179,6 +183,22 @@ fn kafka_source(
+                                        );
+                                    }
+                                }
+                                log.insert(headers_key, Value::from(headers_map));
```
I think you can also just do `headers_map.into()` here.

jszwedko commented on src/sources/kafka.rs:
```diff
@@ -179,6 +183,22 @@ fn kafka_source(
+                                let mut headers_map = BTreeMap::new();
+                                if let Some(headers) = msg.headers() {
+                                    for i in 0..headers.count() {
+                                        let header = headers.get(i).unwrap();
```
Could we make this `.expect("header does not exist")`? I see it only returns `None` if the index is out-of-bounds. This is a bit of an odd API that rdkafka is exposing. I would have expected to see it define `Iter`.

jszwedko commented on src/sources/kafka.rs:
```diff
@@ -179,6 +183,22 @@ fn kafka_source(
+                                        headers_map.insert(
+                                            header.0.to_string(),
+                                            Value::from(
+                                                String::from_utf8_lossy(header.1).to_string(),
```
I might recommend just storing them into `Event` as `Bytes` `Value::from<String>` actually just converts it back to bytes.

https://github.com/timberio/vector/blob/ab0beeeb5b2e7736816a529dce8f18bef8342414/src/event/value.rs#L54-L58

jszwedko reviewed (commented):
Thanks for the contribution @daniilguit !

I left a few comments, but overall this looks good to me.

Would you mind adding it the docs too? You can see https://github.com/timberio/vector/blob/ab0beeeb5b2e7736816a529dce8f18bef8342414/docs/reference/components/sources/kafka.cue#L102-L112 for an example.

daniilguit commented:
@jszwedko: Thank you for your review. Addressed all of your comments but `headers_map.into()` because with this code rust fails to deduce `BTreeMap` type.

bruceg commented on src/sources/kafka.rs:
```diff
@@ -179,6 +183,21 @@ fn kafka_source(
                             }
 
+                            if let Some(headers_key) = &headers_key {
+                                let mut headers_map = BTreeMap::new();
+                                if let Some(headers) = msg.headers() {
+                                    for i in 0..headers.count() {
+                                        let header =
+                                            headers.get(i).expect("header index is out of bounds");
+                                        headers_map.insert(
+                                            header.0.to_string(),
+                                            Bytes::from(header.1.to_owned()).into(),
+                                        );
+                                    }
+                                }
```
This would probably be easier done with iterators and adapters (if `msg.headers()` returns an owned vec and not a reference, used `into_iter()`):
```suggestion
                                let headers_map = msg
                                    .headers()
                                    .map(|headers| {
                                        headers
                                            .iter()
                                            .map(|(key, value)| (key.to_string(), Bytes::from(value.to_owned()).into()))
                                            .collect()
                                    })
                                    .unwrap_or_else(|| BTreeMap::default());
```

jszwedko replied:
Agreed. `.headers()` actually returns a trait object: https://docs.rs/rdkafka/0.26.0/rdkafka/message/trait.Headers.html . I was a little surprised to find this.

daniilguit replied:
@bruceg , @jszwedko  - I can add comment below this `for` loop mentioning this is because `Headers` do not proivde `.iter`.

jszwedko reviewed (approved):
This looks good to me. Thanks @daniilguit !

jszwedko commented:
@daniilguit it looks like the DCO check is failing on the latest commit. Would you be able to `git checkout --amend --sign-off`? See https://github.com/timberio/vector/pull/6815/checks?check_run_id=2159887394

jszwedko commented:
@bruceg I think this is good-to-go if you wouldn't mind giving it another look.

bruceg commented on src/sources/kafka.rs:
```diff
@@ -179,6 +183,23 @@ fn kafka_source(
+                            if let Some(headers_key) = &headers_key {
+                                let mut headers_map = BTreeMap::new();
+                                if let Some(headers) = msg.headers() {
+                                    // Using index-based for loop because rdkafka's `Headers` trait
+                                    // does not provide Iterator-based API
```
Ugh, that's… disappointing, but yes that does require this kind of interface. :disappointed:

bruceg commented on src/sources/kafka.rs:
```diff
@@ -179,6 +183,23 @@ fn kafka_source(
+                                    // Using index-based for loop because rdkafka's `Headers` trait
+                                    // does not provide Iterator-based API
+                                    for i in 0..headers.count() {
+                                        let header =
+                                            headers.get(i).expect("header index is out of bounds");
+                                        headers_map.insert(
+                                            header.0.to_string(),
+                                            Bytes::from(header.1.to_owned()).into(),
+                                        );
```
I think I'd still prefer this to avoid the `.expect()`:
```suggestion
                                        if let Some(header) = headers.get(i) {
                                            headers_map.insert(
                                                header.0.to_string(),
                                                Bytes::from(header.1.to_owned()).into(),
                                            );
                                        }
```

bruceg reviewed (commented):
Looks good, but I'm a little twitchy about the `.expect`.

jszwedko commented:
@bruceg I think this is good now if you want to take one last peek. https://github.com/timberio/vector/pull/6815/commits/7a9e19447ec865b296e66cd2d9275b7c21ec110b removes the `expect`

jszwedko commented:
I brought this up-to-date with master and switched the default behavior to be to add the headers since we updated the rest of the keys to default to being included in https://github.com/timberio/vector/pull/6928

@bruceg do you mind giving another 👀 ?

jszwedko commented:
Closing this in-lieu of https://github.com/timberio/vector/pull/7030 with master merged in. I think this is good-to-go though, thanks for the contribution @daniilguit !
